spec/controllers/user_reports_controller_spec.rb

```diff
@@ -0,0 +1,167 @@
+require "rails_helper"
+
+RSpec.describe UserReportsController, type: :controller do
+  let!(:user){User.create! full_name: "Name", email: "test@mail.com",
+               password: "password", password_confirmation: "password"}
+  let!(:user2){User.create! full_name: "Name 2", email: "test2@mail.com",
+               password: "password", password_confirmation: "password"}
+  let!(:user3){User.create! full_name: "Name 3", email: "test3@mail.com",
+               password: "password", password_confirmation: "password"}
+  let!(:course){Course.create! name: "Course", description: "A course",
+                 start_time: Time.current}
+  let!(:user_course){user.user_courses.create course: course}
+  let!(:user_course2){user2.user_courses.create course: course}
+  let!(:report_1){user.user_reports.create content: "The first report",
+                   date: 2.days.ago, course: course}
+  let!(:report_2){user.user_reports.create content: "The second report",
+                   date: 1.day.ago, course: course}
+  let!(:report_3){user.user_reports.create content: "The third report",
+                   date: Time.now, course: course}
+  let!(:report_4){user2.user_reports.create content: "The first report",
```

Comment:
phân biệt let! và let
k lạm dụng let! nhé e.

Reply:
Vâng.

Reply:
Phân biệt let với let!:
- let: sử dụng lazy evaluation, chỉ có tác dụng trong từng example, chỉ được thực hiện bên trong example nếu như được gọi và được cache giá trị lại trong example đó
- let!: để ép buộc được gọi trước khi vào thực thi từng example, cũng cache lại giống như let

Reply:
example là gì e?

Reply:
- example đối với rspec là một test cơ sở, gồm có đầu vào và đầu ra mong muốn, trong đó với mỗi example chỉ nên kiểm tra đầu ra mong muốn với một đối tượng, được bọc trong lệnh có từ khóa là "it"
- Ngoài ra còn sử dụng các từ khóa describe, context. Trong đó describe dùng để nhóm các example theo class, method,... còn context thì dùng để nhóm các example theo điều kiện đầu vào.